pkg/executor/sortexec/sort.go

```diff
@@ -336,11 +474,112 @@ func (e *SortExec) fetchRowChunks(ctx context.Context) error {
 		return err
 	}
 
-	e.sortPartitions = append(e.sortPartitions, e.curPartition)
+	e.Unparallel.sortPartitions = append(e.Unparallel.sortPartitions, e.curPartition)
 	e.curPartition = nil
 	return nil
 }
 
+func (e *SortExec) fetchChunksParallel(ctx context.Context) error {
+	workerNum := len(e.Parallel.workers)
+
+	// Wait for the finish of all workers
+	workersWaiter := sync.WaitGroup{}
+
+	// Add before the start of goroutine to avoid that the counter is minus to negative.
+	workersWaiter.Add(workerNum + 1)
+
+	// Fetch chunks from child and put chunks into MPMCQueue
+	go e.fetchChunksFromChild(ctx, &workersWaiter)
+
+	// Create and run workers
+	for i := range e.Parallel.workers {
+		e.Parallel.workers[i] = newParallelSortWorker(i, e.lessRow, e.Parallel.globalSortedRowsQueue, &workersWaiter, &e.Parallel.result, e.Parallel.chunkChannel, e.tryToCloseChunkChannel, e.checkErrorForParallel, e.processErrorForParallel, e.memTracker)
+		go e.Parallel.workers[i].run()
+	}
+
+	// Wait for the finish of all workers
+	workersWaiter.Wait()
+	err := e.checkErrorForParallel()
+	if err != nil {
+		return err
+	}
+	e.fetchResultFromQueue()
+	return nil
+}
+
+// Fetch chunks from child and put chunks into MPMCQueue
+func (e *SortExec) fetchChunksFromChild(ctx context.Context, waitGroup *sync.WaitGroup) {
+	defer func() {
+		if r := recover(); r != nil {
+			processPanicAndLog(e.processErrorForParallel, r)
+		}
+		e.tryToCloseChunkChannel()
+		waitGroup.Done()
+	}()
+
+	for {
+		chk := exec.TryNewCacheChunk(e.Children(0))
+		err := exec.Next(ctx, e.Children(0), chk)
+		if err != nil {
+			e.processErrorForParallel(err)
+			return
+		}
+
+		rowCount := chk.NumRows()
+		if rowCount == 0 {
+			break
+		}
+
+		chkWithMemoryUsage := &chunkWithMemoryUsage{
+			Chk:         chk,
+			MemoryUsage: chk.MemoryUsage() + chunk.RowSize*int64(rowCount),
+		}
+
+		e.memTracker.Consume(chkWithMemoryUsage.MemoryUsage)
+
+		// chunkChannel may be closed by workers and it's ok to let it panic
+		// as workers have panicked and the query can't keep on running.
+		e.Parallel.chunkChannel <- chkWithMemoryUsage
+
+		err = e.checkErrorForParallel()
+		if err != nil {
+			return
+		}
+
+		injectParallelSortRandomFail()
+	}
+}
+
+func (e *SortExec) checkErrorForParallel() error {
+	e.Parallel.errRWLock.RLock()
+	defer e.Parallel.errRWLock.RUnlock()
+	return e.Parallel.err
+}
+
+func (e *SortExec) processErrorForParallel(err error) {
+	e.Parallel.errRWLock.Lock()
+	defer e.Parallel.errRWLock.Unlock()
+	if e.Parallel.err == nil {
+		e.Parallel.err = err
+	}
+}
+
+func (e *SortExec) fetchResultFromQueue() {
+	sortedRowsNum := e.Parallel.globalSortedRowsQueue.getSortedRowsNumNoLock()
+	if sortedRowsNum > 1 {
+		panic("Sort is not completed.")
+	}
+
```

Comment:
```suggestion

	idx = 0
```

Reply:
> 

Why we need to add `idx = 0`?

Reply:
In golang, is the default value of int 0? If so, there is no need to call `idx = 0` here.

Reply:
> In golang, is the default value of int 0? If so, there is no need to call `idx = 0` here.

Yes, the default value of int in golang is 0.